Support custom initial options for sqlite

ghassmo opened the pull request:
Hello, 
This addresses issue #621, the key (and any options) for sqlcipher must be set before the first operation on the database. the page_size option is set before setting any custom options.

abonander commented on sqlx-core/src/sqlite/options/mod.rs:
```diff
@@ -183,6 +185,14 @@ impl SqliteConnectOptions {
+    /// Sets custom initial settings for the database connection.
+    ///
+    /// The default initial_options is NONE.
+    pub fn initial_options(mut self, initial_options: &'static str) -> Self {
```
The documentation and name for this method is really vague. It sounds like a list of `PRAGMAS` to specify but really it's a whole statement to execute.

For clarity's sake, why not give it the name `pragmas` and have it take a list of key-value pairs? Collect them internally to a vector or order-preserving map and deduplicate the keys with those we set by default.

abonander replied:
A `&'static str` isn't that useful here anyway, especially if you're setting a SQLCipher encryption key (which would normally be derived at runtime from a user-entered password).

ghassmo replied:
Yes you are correct, i will change it.

abonander commented on sqlx-core/src/sqlite/options/connect.rs:
```diff
@@ -18,14 +18,23 @@ impl ConnectOptions for SqliteConnectOptions {
+            conn.execute(&*page_size_opt).await?;
+
+            // apply custom initial_options
+            match self.initial_options {
+                Some(stmt) => {
+                    conn.execute(&*stmt).await?;
+                }
+                None => (),
+            }
```
Unless you intended to add something to the `None` branch:

```suggestion
            if let Some(stmt) = self.initial_options {
                    conn.execute(&*stmt).await?;
            }
```

ghassmo commented:
@abonander  as you suggested, I added a new field "pragams" of type map inside SqliteConnectOptions and it iterate over that map when send initial sql statements, also i added pragma function which add custom new key and value to that map

to preserve the order inside the map and avoid dublicate keys i am using indexmap crate

abonander commented on sqlx-core/src/sqlite/options/mod.rs:
```diff
@@ -53,16 +55,11 @@ pub struct SqliteConnectOptions {
-    pub(crate) synchronous: SqliteSynchronous,
-    pub(crate) auto_vacuum: SqliteAutoVacuum,
-    pub(crate) page_size: u32,
+    pub(crate) pragmas: IndexMap<String, String>,
```
You could save some allocations by using `Cow<'static, str>` here instead of `String`

abonander commented on sqlx-core/src/sqlite/options/connect.rs:
```diff
@@ -18,22 +18,9 @@ impl ConnectOptions for SqliteConnectOptions {
-            conn.execute(&*init).await?;
+            for (key, value) in self.pragmas.iter() {
+                conn.execute(&*format!("PRAGMA {} = {};", key, value)).await?;
+            }
```
Since there is a bit of overhead in `.execute()` I think it would be better if we collected the `PRAGMA` statements together in one string and executed that.

abonander commented on sqlx-core/src/sqlite/options/mod.rs:
```diff
@@ -171,23 +195,31 @@ impl SqliteConnectOptions {
+    }
+
+    /// Sets custom initial pragmas for the database connection.
+    pub fn pragma(mut self, key: &str, value: &str) -> Self {
```
If you make the `Cow<'static, str>` change above, I would make `key` and `value` both take `impl Into<Cow<'static, str>>` here.

abonander reviewed (changes requested):
A few nits but otherwise it looks good!

abonander commented on sqlx-core/src/sqlite/options/connect.rs:
```diff
@@ -18,20 +18,11 @@ impl ConnectOptions for SqliteConnectOptions {
+            let mut init = String::new();
+
+            for (key, value) in self.pragmas.iter() {
+                init += &format!("PRAGMA {} = {}; ", key, value);
```
Add an import for `std::fmt::Write`, then:

```suggestion
               write!(init, "PRAGMA {} = {}; ", key, value).ok();
```

abonander commented on sqlx-core/src/sqlite/options/mod.rs:
```diff
@@ -171,23 +195,35 @@ impl SqliteConnectOptions {
+    }
+
+    /// Sets custom initial pragma for the database connection.
+    pub fn pragma<T>(mut self, key: T, value: T) -> Self
+    where
+        T: Into<Cow<'static, str>>,
```
```suggestion
    pub fn pragma<K, V>(mut self, key: K, value: V) -> Self
    where
        K: Into<Cow<'static, str>>,
        V: Into<Cow<'static, str>>,
```

abonander reviewed (changes requested):
Sorry, just a couple more nits and this should be good to go.

abonander commented on sqlx-core/src/sqlite/options/mod.rs:
```diff
@@ -53,17 +55,11 @@ pub struct SqliteConnectOptions {
-    pub(crate) auto_vacuum: SqliteAutoVacuum,
-    pub(crate) page_size: u32,
-    pub(crate) immutable: bool,
+    pub(crate) pragmas: IndexMap<Cow<'static, str>, Cow<'static, str>>,
```
```suggestion
    pub(crate) immutable: bool,
    pub(crate) pragmas: IndexMap<Cow<'static, str>, Cow<'static, str>>,
```

abonander commented on sqlx-core/src/sqlite/options/connect.rs:
```diff
@@ -18,20 +18,11 @@ impl ConnectOptions for SqliteConnectOptions {
+            let mut init = String::new();
+
+            for (key, value) in self.pragmas.iter() {
+               write!(init, "PRAGMA {} = {}; ", key, value).ok();
```
```suggestion
               use std::fmt::Write;
               write!(init, "PRAGMA {} = {}; ", key, value).ok();
```

abonander replied:
```suggestion
                use std::fmt::Write;
                write!(init, "PRAGMA {} = {}; ", key, value).ok();
```

abonander commented on sqlx-core/src/sqlite/options/mod.rs:
```diff
@@ -74,22 +71,43 @@ impl Default for SqliteConnectOptions {
-            auto_vacuum: Default::default(),
-            page_size: 4096,
-            immutable: false,
+            pragmas,
```
```suggestion
            immutable: false,
            pragmas,
```

abonander commented:
Sorry, I'm making a mess of this PR by trying to get it passing in CI using the "Suggest a Change" feature in Github, it's very error-prone. All these excess commits should be removed by the squash.

abonander commented on sqlx-core/src/sqlite/options/connect.rs:
```diff
@@ -18,20 +18,13 @@ impl ConnectOptions for SqliteConnectOptions {
+            let mut init = String::new();
+
+            for (key, value) in self.pragmas.iter() {
+               use std::fmt::Write;
```
```suggestion

```

joenano commented:
Is there any guide or examples on how to connect to a sqlcipher encrypted database?

reivilibre commented:
A bit late to the party but https://github.com/launchbadge/sqlx/blob/e783f9d47c6779304aa569384cabccd468b63e68/tests/sqlite/sqlcipher.rs#L52-L56 seems to be an example of using sqlcipher ­— I haven't tested it yet. I was confused how this PR would actually let you change the first pragma because it still sets a bunch of default pragmas in the `new` method and then the pragmas are written out in-order.

But actually, I don't think this PR *was* sufficient, look at what was added to the code since:
https://github.com/launchbadge/sqlx/blob/e783f9d47c6779304aa569384cabccd468b63e68/sqlx-sqlite/src/options/mod.rs#L118-L119
(There is explicitly a `None` entry added for the key first in the pragma map.)

So it should be sufficient to call `.pragma("key", "hunter2")`, without special care worrying about the order or anything, on the `SqliteConnectionOptions` because the place in the map was already reserved by SQLx.
